README.md
# Konami Code For jQuery
By [Tom McFarlin](https://tommcfarlin.com). Last Updated 4 December 2015.

## About

Using the Konami code, easily configure and Easter Egg for your page or any element on the page.

## Parameters

* `code` Personalized code.
* `cheat` The callback function to fire once the cheat code has been entered.
* `eventName` jQuery event name for default callback
* `eventProperties` event property override for default callback

## Installation
```sh
bower install konami-code
```

## Instructions

Include the plugin in the header of your page:

```html
	<script type="text/javascript" src="//ajax.googleapis.com/ajax/libs/jquery/1.10.2/jquery.min.js"></script>`
	<script src="jquery.konami.js" type="text/javascript"></script>
```

### With callback

Apply the plugin to a selector to capture keypresses:

```js
  $( window ).konami();
```

```js
  $( '.konami-sensitive' ).konami();
```

Specify a callback to fire once the code has been entered:

```js
  $( window ).konami({
		cheat: function() {
			alert( 'Cheat code activated!' );
		}
	});
```

### Using jQuery events

Catch the konami code with a jQuery event handler:

```js
  $( window ).konami();
  $( window ).on('konami', function() {
    alert( 'Cheat code activated!' );
  })
```

Add extra data to the jQuery event callback:

```js
  $( window ).konami( { message: 'special message' } );
  $( window ).on('konami', function(evt, extraParam) {
    alert( 'Cheat code activated: ' + extraParam.message + '!' );
  })
```

Use event names:

```js
  $('.type1').konami( { eventName: 'konami.on.type1' } );
  $('.type2').konami( { eventName: 'konami.on.type2' } );
  $( window ).on('konami.on.type2', function(evt, extraParam) {
    alert( 'Cheat code activated on a type2 element' );
  })
```

### Personallizing the code

You can personalize the code too, just entering a array with ASCII codes keys in code param

```js
  $( window ).konami({
  		code : [38,38,40,40,37,39,37,39], // up up down down left right left right
		cheat: function() {
			alert( 'Cheat code activated!' );
		}
	});
```


## Contact

* Web: [Tom McFarlin](https://tommcfarlin.com)
* Twitter: [@tommcfarlin](https://twitter.com/tommcfarlin/)

## License

[MIT license](http://www.opensource.org/licenses/mit-license.php)


changelog.md
_1.3.3_ (4 December 2015)

* Updating README
* Tagging for Bower support

_1.3.1_

* jQuery events supported

_1.3.0 (7 March 2014)_

* Minor updates to the coding conventions and docblocks
* Tagging for registration in the Bower repository

_1.2.2 (6 March 2014)_

* Added support for bower (bower.json)

_1.2.1 (23 October 2013)_

* Added option to customize the code

_1.2.0 (11 October 2013)_

* Additional, minor refactoring for strong JSLint
* Removing extranneous whitespace
* Moving the plugin to a `src` directory
* Updated the demo to the latest version of jQuery
* Updated the code for jQuery
* Updated authorship information
* Updated the version number in the plugin
* Updated the README and the changelog

_1.1.0 (10, October, 2013)_

* Refactoring some of the code up to JSLint standards
* Updating the changelog to markdown
* Updating the README to markdown
* Removed the MIT License and included it in the README

_1.0.0 4 March 2011_

* Initial release
* 3/14/2011

src/vanilla.konami.js
/*
 * Konami Code Javascript Object
 * 1.3.0, 7 March 2014
 *
 * Using the Konami code, easily configure an Easter Egg for your page or any element on the page.
 *
 * Options:
 * - code : set your own custom code, takes array of keycodes / default is original Konami code
 * - cheat : the function to call when the proper sequence is entered
 * - elem : the element to set the instance on
 *
 * Copyright 2013 - 2014 Kurtis Kemple, http://kurtiskemple.com
 * Released under the MIT License
 */

var KONAMI = function ( options ) {
	var elem, ret, defaults, keycode, config, cache;

	// set the default code,function, and element
	defaults = {
		code : [38,38,40,40,37,39,37,39,66,65],
		cheat : null,
		elem : window
	};

	// build our return object
	ret = {

		/**
		 * handles the initialization of the KONAMI instance
		 *
		 * @param  {object} options the config to pass in to the instance
		 * @return {none}
		 * @method  init
		 * @public
		 */
		init : function ( options ) {
			cache = [], config = {};

			if ( options ) {

				for ( var key in defaults ) {

					if ( defaults.hasOwnProperty( key ) ) {

						if ( !options[ key ] ) {

							config[ key ] = defaults[ key ];
						} else {

							config[ key ] = options[ key ];
						}
					}
				}
			} else {

				config = defaults;
			}

			ret.bind( config.elem, 'keyup', ret.konami );
		},

		/**
		 * handles disassembling of the instance
		 *
		 * @return {none}
		 * @method  destroy
		 * @public
		 */
		destroy : function () {
			ret.unbind( config.elem, 'keyup', ret.konami );
			cache = config = null;
		},

		/**
		 * handles adding events to elements
		 *
		 * @param   {elem}     elem  DOM element to attach to
		 * @param   {string}   evt   the event type to bind to
		 * @param   {Function} fn    the function to bind
		 * @return  {none}
		 * @method  bind
		 * @private
		 */
		bind : function ( elem, evt, fn ) {
			if ( elem.addEventListener ) {

				elem.addEventListener( evt, fn, false );
			} else if ( elem.attachEvent ) {

				elem.attachEvent( 'on'+ evt, function( e ) {
					fn( e || window.event );
				});
			}
		},

		/**
		 * handles removing events from elements
		 *
		 * @param   {elem}     elem DOM element to remove from
		 * @param   {string}   evt  the event type to unbind
		 * @param   {Function} fn   the function to unbind
		 * @return  {none}
		 * @method  unbind
		 * @private
		 */
		unbind : function ( elem, evt, fn ) {
			if ( elem.removeEventListener ) {

				elem.removeEventListener( evt, fn, false );
			} else if ( elem.detachEvent ) {

				elem.detachEvent( 'on' + evt, function( e ) {
					fn( e || window.event );
				});
			}
		},

		/**
		 * handles the business logic for checking for valid konami code
		 *
		 * @param   {object} e the event object
		 * @return  {none}
		 * @method  konami
		 * @private
		 */
		konami : function( e ) {
			keycode = e.keyCode || e.which;

			if ( config.code.length > cache.push( keycode ) ) {

				return;
			}

			if ( config.code.length < cache.length ) {

				cache.shift();
			}

			if ( config.code.toString() !== cache.toString() ) {

				return;
			}

			config.cheat();
		}
	};

	ret.init( options );
	return ret;
};

/*=========================================
	Example Usage:

	var options = {
		cheat : function() {
			alert( 'KONAMI!!' );
		}
	};

	var example = new KONAMI( options );

		// rebuild with new settings
		example.destroy();
		example.init( newConfig );
==========================================*/

src/jquery.konami.js
/*
 * Konami Code For jQuery Plugin
 * 1.3.3, 4 December 2015
 *
 * Using the Konami code, easily configure and Easter Egg for your page or any element on the page.
 *
 * Copyright 2011 - 2015 Tom McFarlin, https://tommcfarlin.com
 * Released under the MIT License
 */

(function ( $ ) {
	"use strict";

	$.fn.konami = function( options ) {
		var opts, controllerCode;

		opts = $.extend({}, $.fn.konami.defaults, options);
		controllerCode = [];

		// Note that we use the passed-in options, not the resolved options.
		opts.eventProperties = $.extend({}, options,  opts.eventProperties);

		this.keyup(function( evt ) {
			var code = evt.keyCode || evt.which;

			if ( opts.code.length > controllerCode.push( code ) ) {
				return;
			}

			if ( opts.code.length < controllerCode.length ) {
				controllerCode.shift();
			}

			if ( opts.code.toString() !== controllerCode.toString() ) {
				return;
			}

			opts.cheat(evt, opts);

		});

		return this;
	};

	$.fn.konami.defaults = {
		code : [38,38,40,40,37,39,37,39,66,65],
		eventName : 'konami',
		eventProperties : null,
		cheat: function(evt, opts) {
			$(evt.target).trigger(opts.eventName, [ opts.eventProperties ]);
		}
	};

})( jQuery );